### PR TITLE
Adding a new transport class to handle Phoenix channels

### PR DESCRIPTION
**Pull request by acatinon**
First of all, thank you for this great library !

I'm building a tool that connect to a server built on [Absinthe](http://absinthe-graphql.org/). This toolkit relies on the [Phoenix](https://www.phoenixframework.org/) framework [channels](https://hexdocs.pm/phoenix/Phoenix.Channel.html#content) feature to handle GraphQL subscriptions.

So I created a `PhoenixChannelWebsocketsTransport` class that implement all the logic to subscribe and receive data from a Phoenix channel.

Of course I'm open to any suggestions to improve this PR

### CODE REVIEW

**Comment by KingDarBoja** [comment]
Not an expert on Elixir related framework/libraries but this transport class could be added as [extra optional feature](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) so any Elixir (Phoenix in this case) user can install it by running the following command:

```shell
pip install gql[phoenix-transport]
```

@leszekhanusz what do you think?

Also, I will take a look at it whenever I can, currently busy with some `graphql-server-core` stuff.

**Comment by acatinon** [comment]
> I tried to find a spec which describe the protocol implemented here but I have not found it. Could you please add a link to it ?

@leszekhanusz Unfortunately I didn't find any specs for Phoenix channels either. There are  pointers to some docs in [this thread](https://elixirforum.com/t/specification-of-the-protocol-used-by-phoenix-channels/4192/3), but i've mainly done reverse engineering on the [C# Phoenix client](https://github.com/Mazyod/PhoenixSharp) project.

Btw I will take your feedbacks into account.

**Comment by acatinon** [comment]
@leszekhanusz I finished taking your comments into account.

**Comment by acatinon** [comment]
Do you need me to further improve the test coverage on this PR ?

**Comment by KingDarBoja** [comment]
> Do you need me to further improve the test coverage on this PR ?

The `/gql/transport/phoenix_channel_websockets.py` has 96.26% of coverage as seen at [coveralls report](https://coveralls.io/jobs/64224005/source_files/975200836), but I think it is pretty high already.

So don't worry about it at the moment.

**Comment by KingDarBoja** [comment]
@acatinon updating with the latest changes from master should bring coverage to 100%, also there is few mypy errors on the phoenxi transport.

**Comment by acatinon** [comment]
@KingDarBoja Done.

File: gql/transport/phoenix_channel_websockets.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        self.next_query_id += 1
+
+        init_message = json.dumps(
+            {
+                "topic": self.channel_name,
+                "event": "phx_join",
+                "payload": {},
+                "ref": query_id,
+            }
+        )
```
Comment (KingDarBoja) [inline gql/transport/phoenix_channel_websockets.py:41]: My general though would be providing a [class typing](https://www.python.org/dev/peps/pep-0589/) for the json events as those are documented at [websocket-clients-and-phoenix-channels](http://graemehill.ca/websocket-clients-and-phoenix-channels/) made by [Graeme](https://elixirforum.com/t/specification-of-the-protocol-used-by-phoenix-channels/4192/7) on the linked Elixir Topic.

Unfortunately, this Python feature is available at Python 3.8+ as stated on the (library docs)(https://docs.python.org/3/library/typing.html#typing.TypedDict). In order to keep it backward compatible, the mypy suggest using [typing-extensions](https://mypy.readthedocs.io/en/stable/more_types.html#typeddict).

@leszekhanusz thoughts?

File: gql/transport/phoenix_channel_websockets.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+
+class PhoenixChannelWebsocketsTransport(WebsocketsTransport):
+    def __init__(
+        self, channel_name: str, heartbeat_interval: int = 30, *args, **kwargs
+    ) -> None:
+        self.channel_name = channel_name
+        self.heartbeat_interval = heartbeat_interval
+        self.subscription_ids_to_query_ids: Dict[str, int] = {}
+        super(PhoenixChannelWebsocketsTransport, self).__init__(*args, **kwargs)
```
Comment (KingDarBoja) [inline gql/transport/phoenix_channel_websockets.py:23]: Would be great to document where these comes from. It seems like [Phoenix.Socket](https://hexdocs.pm/phoenix/Phoenix.Socket.html) and [Phoenix.Socket.Transport](https://hexdocs.pm/phoenix/Phoenix.Socket.Transport.html) are the keys for this task.

File: gql/transport/phoenix_channel_websockets.py (hunk 1 above)
Comment (leszekhanusz) [reply gql/transport/phoenix_channel_websockets.py:41]: IMO the typing hints are particularly interesting for prototype functions where they could be seen as a kind of documentation about the function arguments. I don't see the need to add them necessarily to all variables.

And in this case we don't really do that in the previous websockets.py code so I don't see why we should add it here.

File: gql/transport/phoenix_channel_websockets.py (hunk 2 above)
Comment (acatinon) [reply gql/transport/phoenix_channel_websockets.py]: Done.

**Comment by KingDarBoja** [comment]
This one is almost ready to merge, only thing left is increasing coverage to 100%. 

cc @leszekhanusz

**Review by KingDarBoja — APPROVED** [review]
Coverage reached 100% 🚀

**Comment by leszekhanusz** [comment]
> Coverage reached 100% rocket

I cheated a little and just added pragmas for the few remaining lines...

**Review by leszekhanusz — APPROVED** [review]

**Comment by leszekhanusz** [comment]
@acatinon Thanks again for this PR and sorry for the delay

**Pull request merged**